src/c++/perf_analyzer/test_concurrency_manager.cc

```diff
@@ -0,0 +1,306 @@
+    } else {
+      return ConcurrencyManager::Infer(thread_stat, thread_config);
+    }
+  }
+
+  // Mock out most of the complicated Infer code
+  //
+  void MockInfer(
+      std::shared_ptr<ThreadStat> thread_stat,
+      std::shared_ptr<ThreadConfig> thread_config)
+  {
+    if (!execute_) {
+      thread_config->is_paused_ = true;
+    }
+  }
+
+  /// Test that the correct Infer function is called in the backend
+  ///
+  void TestInferType()
+  {
+    // FIXME TMA-982: This delay is to avoid deadlock. Investigate why delay is
+    // needed.
+    stats_->request_delay = std::chrono::milliseconds(50);
+
+    ChangeConcurrencyLevel(params_.max_concurrency);
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(500));
+
+    CheckInferType();
+  }
+
+  /// Test that the correct concurrency is maintained in the load manager
+  ///
+  void TestConcurrency(
+      std::chrono::milliseconds request_delay,
+      std::chrono::milliseconds sleep_time)
+  {
+    stats_->request_delay = request_delay;
+
+    ChangeConcurrencyLevel(params_.max_concurrency);
+    std::this_thread::sleep_for(sleep_time);
+
+    CheckConcurrency();
+  }
+
+  /// Test sequence handling
+  ///
+  void TestSequences()
+  {
+    auto sleep_time = std::chrono::milliseconds(500);
+
+    ChangeConcurrencyLevel(params_.max_concurrency);
+    std::this_thread::sleep_for(sleep_time);
+
+    CheckSequences();
+  }
+
+ private:
+  void CheckConcurrency()
+  {
+    if (params_.max_concurrency < 4) {
+      CHECK(stats_->num_active_infer_calls.load() == params_.max_concurrency);
+    } else {
+      CHECK(
+          stats_->num_active_infer_calls.load() ==
+          doctest::Approx(params_.max_concurrency).epsilon(0.25));
+    }
+  }
+
+  const bool use_mock_infer_{false};
+};
+
+/// Test that the correct Infer function is called in the backend
+///
+TEST_CASE("concurrency_infer_type")
+{
+  PerfAnalyzerParameters params{};
+  params.max_concurrency = 1;
+
+  SUBCASE("async_streaming")
+  {
+    params.async = true;
+    params.streaming = true;
+  }
+  SUBCASE("async_no_streaming")
+  {
+    params.async = true;
+    params.streaming = false;
+  }
+  SUBCASE("no_async_streaming")
+  {
+    params.async = false;
+    params.streaming = true;
+  }
+  SUBCASE("no_async_no_streaming")
+  {
+    params.async = false;
+    params.streaming = false;
+  }
+
+  TestConcurrencyManager tcm(params);
+  tcm.TestInferType();
+}
+
+/// Test that the correct concurrency is maintained in the load manager
+///
+TEST_CASE("concurrency_concurrency")
+{
+  PerfAnalyzerParameters params{};
+  std::chrono::milliseconds request_delay{50};
+  std::chrono::milliseconds sleep_time{225};
+
+  SUBCASE("sync, no-streaming, 1 concurrency, 1 thread")
+  {
+    params.forced_sync = true;
+    params.async = false;
+    params.streaming = false;
+    params.max_concurrency = 1;
+    params.max_threads = 1;
+  }
+
+  SUBCASE("sync, no-streaming, 4 concurrency, 4 threads")
```

Comment:
I don't see sync tests where concurrency != threads. Is that a legal combination?

Reply:
PA forces the number of threads to be equal to the concurrency. That's why there's no tests for concurrency != threads (invalid).

https://github.com/triton-inference-server/client/blob/main/src/c%2B%2B/perf_analyzer/command_line_parser.cc#L337-L340

```cpp
             "However, when running"
             "in synchronous mode with concurrency-range having explicit 'end' "
             "specification,"
             "this value will be ignored. Default is 4 if --request-rate-range "
```

https://github.com/triton-inference-server/client/blob/main/src/c%2B%2B/perf_analyzer/perf_analyzer.cc#L190-L191

```cpp
        params_->max_threads = std::max(
            params_->concurrency_range.start, params_->concurrency_range.end);
```

https://github.com/triton-inference-server/client/blob/main/src/c%2B%2B/perf_analyzer/concurrency_manager.cc#L104-L122

```cpp
  while ((concurrent_request_count > threads_.size()) &&
         (threads_.size() < max_threads_)) {
...
    threads_.emplace_back(
        &ConcurrencyManager::Infer, this, threads_stat_.back(),
        threads_config_.back());
  }
```